From efb2d37aebe3690bdf8153ea7f8a0772e9dcb18c Mon Sep 17 00:00:00 2001
From: "GC.Pelosi" <giovanni.pelosi@unimib.it>
Date: Thu, 24 Sep 2020 10:33:41 +0200
Subject: [PATCH] if class (-c) strart with 'St-', load 'St' as default

    see: /etc/X11/app-defaults/XTerm*
---
 x.c | 38 +++++++++++++++++++++++++++++---------
 1 file changed, 29 insertions(+), 9 deletions(-)

diff --git a/x.c b/x.c
index 76f167f..9a56d91 100644
--- a/x.c
+++ b/x.c
@@ -1977,23 +1977,15 @@ run(void)
 }
 
 int
-resource_load(XrmDatabase db, char *name, enum resource_type rtype, void *dst)
+resource_read(XrmDatabase db, char *name, enum resource_type rtype, void *dst, char *fullname, char *fullclass)
 {
 	char **sdst = dst;
 	int *idst = dst;
 	float *fdst = dst;
 
-	char fullname[256];
-	char fullclass[256];
 	char *type;
 	XrmValue ret;
 
-	snprintf(fullname, sizeof(fullname), "%s.%s",
-			opt_name ? opt_name : "st", name);
-	snprintf(fullclass, sizeof(fullclass), "%s.%s",
-			opt_class ? opt_class : "St", name);
-	fullname[sizeof(fullname) - 1] = fullclass[sizeof(fullclass) - 1] = '\0';
-
 	XrmGetResource(db, fullname, fullclass, &type, &ret);
 	if (ret.addr == NULL || strncmp("String", type, 64))
 		return 1;
@@ -2012,6 +2004,34 @@ resource_load(XrmDatabase db, char *name, enum resource_type rtype, void *dst)
 	return 0;
 }
 
+int
+resource_load(XrmDatabase db, char *name, enum resource_type rtype, void *dst)
+{
+	char fullname[256];
+	char fullclass[256];
+	char baseclass[256];
+
+        int rc;
+
+        /* if class starts with 'St-', load default value for class 'St' */
+
+	snprintf(fullname, sizeof(fullname), "%s.%s",
+			opt_name ? opt_name : "st", name);
+	snprintf(fullclass, sizeof(fullclass), "%s.%s",
+			opt_class ? opt_class : "St", name);
+	fullname[sizeof(fullname) - 1] = fullclass[sizeof(fullclass) - 1] = '\0';
+
+	if (opt_class && !strncmp("St-", opt_class, 3)) { 
+          snprintf(baseclass, sizeof(baseclass), "%s.%s",
+			"St", name);
+          baseclass[sizeof(baseclass) - 1] = '\0';
+          rc = resource_read(db,name,rtype,dst, fullname, baseclass);
+        }
+
+        rc = resource_read(db,name,rtype,dst, fullname, fullclass);
+	return rc;
+}
+
 void
 config_init(void)
 {
-- 
2.7.4

